hashing/src/chunk_with_proof.rs

```diff
@@ -0,0 +1,307 @@
+use crate::indexed_merkle_proof::IndexedMerkleProof;
+
+use casper_types::{
+    allocate_buffer,
+    bytesrepr::{Bytes, FromBytes, ToBytes},
+};
+
+use crate::{error, Digest};
+
+#[derive(PartialEq, Debug, schemars::JsonSchema, serde::Serialize, serde::Deserialize)]
+#[schemars(with = "String", description = "Hex-encoded hash digest.")]
+#[serde(deny_unknown_fields)]
+pub struct ChunkWithProof {
+    proof: IndexedMerkleProof,
+    chunk: Vec<u8>,
+}
+
+impl ToBytes for ChunkWithProof {
+    fn to_bytes(&self) -> Result<Vec<u8>, casper_types::bytesrepr::Error> {
+        let mut result = allocate_buffer(self)?;
+        result.append(&mut self.proof.to_bytes()?);
+        result.append(&mut Bytes::from(self.chunk.as_slice()).to_bytes()?);
+        Ok(result)
+    }
+
+    fn serialized_length(&self) -> usize {
+        self.proof.serialized_length() + self.chunk.serialized_length()
+    }
+}
+
+impl FromBytes for ChunkWithProof {
+    fn from_bytes(bytes: &[u8]) -> Result<(Self, &[u8]), casper_types::bytesrepr::Error> {
+        let ((proof, chunk), remainder) = <(IndexedMerkleProof, Bytes)>::from_bytes(bytes)?;
+        Ok((
+            ChunkWithProof {
+                proof,
+                chunk: chunk.into(),
+            },
+            remainder,
+        ))
+    }
+}
+
+impl ChunkWithProof {
+    #[cfg(test)]
+    pub const CHUNK_SIZE_BYTES: usize = 10;
+
+    #[cfg(not(test))]
+    pub const CHUNK_SIZE_BYTES: usize = 1_048_576; // 2^20
+
+    #[allow(unused)]
+    fn new(data: &[u8], index: u64) -> Result<Self, error::MerkleConstructionError> {
+        if data.len() < Self::CHUNK_SIZE_BYTES * (index as usize) {
+            return Err(error::MerkleConstructionError::IndexOutOfBounds {
+                count: data.chunks(Self::CHUNK_SIZE_BYTES).len() as u64,
+                index,
+            });
+        }
+
+        let (proof, chunk) = if data.is_empty() {
+            (IndexedMerkleProof::new([Digest::hash(&[])], index)?, vec![])
```

Comment:
Maybe that decision was made because `ChunkWithProof` always has a chunk, so no `ChunkWithProof` could represent empty data if it were represented as "zero chunks" instead of "one empty chunk". That's why `verify` will return false if the proof is empty. Makes sense to me now, but we should document this.

Maybe add to the `ChunkWithProof` documentation that empty data is always represented as _one empty chunk_ and not as _zero chunks_.

Reply:
Added doc comment for `ChunkWithProof` struct.

Reply:
Do we want an explicit test for this?